optuna/storages/cached_storage.py

```diff
@@ -0,0 +1,328 @@
+import copy
+import threading
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Tuple
+
+from optuna import distributions
+from optuna.storages import base
+from optuna.storages.rdb.storage import RDBStorage
+from optuna.study import StudyDirection
+from optuna.study import StudySummary
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+class _TrialUpdates:
+    def __init__(self) -> None:
+        self.state = None  # type: Optional[TrialState]
+        self.value = None  # type: Optional[float]
+        self.values = dict()  # type: Dict[int, float]
+        self.user_attrs = dict()  # type: Dict[str, Any]
+        self.system_attrs = dict()  # type: Dict[str, Any]
+        self.params = dict()  # type: Dict[str, Any]
+        self.distributions = dict()  # type: Dict[str, distributions.BaseDistribution]
+
+
+class _StudyInfo:
+    def __init__(self) -> None:
+        self.trials = dict()  # type: Dict[int, FrozenTrial]
+        self.updates = dict()  # type: Dict[int, _TrialUpdates]
+        self.param_distribution = {}  # type: Dict[str, distributions.BaseDistribution]
+
+
+class _CachedStorage(base.BaseStorage):
+    """A wrapper class of storage backends.
+
+    This class is used in :func:`~optuna.get_storage` function and automatically
+    wraps :class:`~optuna.storages.RDBStorage` class.
+
+    Args:
+        storage:
+            BaseStorage class instance to wrap.
+    """
+
+    def __init__(self, backend: RDBStorage) -> None:
+        self._backend = backend
+        self._studies = {}  # type: Dict[int, _StudyInfo]
+        self._trial_id_to_study_id_and_number = dict()  # type: Dict[int, Tuple[int, int]]
+        self._lock = threading.RLock()
+
+    def __getstate__(self):
+        # type: () -> Dict[Any, Any]
```

Comment:
A nitpicking but let me suggest the following change:
```suggestion
    def __getstate__(self) -> Dict[Any, Any]:
```